### PR TITLE
[image_picker] ios:  support GIF animation and the scaling

### PR DESCRIPTION
**Pull request by ko2ic**
## Description

Follow up https://github.com/flutter/plugins/pull/1617. GIF animation and its scaling were supported.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] My PR includes unit or integration tests for *all* changed/updated/fixed behaviors (See [Contributor Guide]).
- [x] All existing and new tests are passing.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] The analyzer (`flutter analyze`) does not report any problems on my PR.
- [x] I read and followed the [Flutter Style Guide].
- [x] The title of the PR starts with the name of the plugin surrounded by square brackets, e.g. [shared_preferences]
- [x] I updated pubspec.yaml with an appropriate new version according to the [pub versioning philosophy].
- [x] I updated CHANGELOG.md to add a description of the change.
- [x] I signed the [CLA].
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require plugin users to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (please indicate a breaking change in CHANGELOG.md and increment major revision).
- [x] No, this is *not* a breaking change.

### CODE REVIEW

**Comment by ko2ic** [comment]
@cyanglaz I was able to handle GIF animation and its scaling.
Could you review this?

**Comment by ko2ic** [comment]
I pushed again because there was extra processing.

File: packages/image_picker/ios/Classes/ImagePickerPlugin.m [hunk 1]
```diff
@@ -263,7 +264,20 @@ - (void)imagePickerController:(UIImagePickerController *)picker
                            options:nil
                      resultHandler:^(NSData *_Nullable imageData, NSString *_Nullable dataUTI,
                                      UIImageOrientation orientation, NSDictionary *_Nullable info) {
-                       [weakSelf saveImageWithOriginalImageData:imageData image:image];
+                       FLTImagePickerMIMEType type = kFLTImagePickerMIMETypeDefault;
```
Comment (cyanglaz) [inline packages/image_picker/ios/Classes/ImagePickerPlugin.m]: we can move `mage = [FLTImagePickerImageUtil scaledImage:image maxWidth:maxWidth maxHeight:maxHeight];` inside the block just as we did for the gifs. Also, everything in the block might be able to move to a separate method like `[weakSelf saveImageWithOriginalImageData:data image:image type:type maxWidth:width maxHeight:height];`

File: packages/image_picker/ios/Classes/FLTImagePickerImageUtil.m [hunk 2]
```diff
@@ -0,0 +1,106 @@
+// Copyright 2019 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#import "FLTImagePickerImageUtil.h"
+#import <MobileCoreServices/MobileCoreServices.h>
```
Comment (cyanglaz) [inline packages/image_picker/ios/Classes/FLTImagePickerImageUtil.m]: what requires `MobileCoreServices`?

Comment (ko2ic) [reply packages/image_picker/ios/Classes/FLTImagePickerImageUtil.m]: It is necessary to use ```kUTTypeGIF``` on line 65.

File: packages/image_picker/ios/Classes/FLTImagePickerImageUtil.m [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+    NSDictionary *properties = (NSDictionary *)CFBridgingRelease(
+        CGImageSourceCopyPropertiesAtIndex(imageSource, index, NULL));
+    NSDictionary *gifProperties = properties[(NSString *)kCGImagePropertyGIFDictionary];
+
+    NSNumber *delay = gifProperties[(NSString *)kCGImagePropertyGIFUnclampedDelayTime];
+    if (!delay) {
+      delay = gifProperties[(NSString *)kCGImagePropertyGIFDelayTime];
+    }
+
+    duration = [delay doubleValue];
```
[the remark this answers is not in the document]
Comment (ko2ic) [reply packages/image_picker/ios/Classes/FLTImagePickerImageUtil.m]: It is same for any value in the array.
So I only set if there is no value.

File: packages/image_picker/ios/Classes/ImagePickerPlugin.m (hunk 1 above)
Comment (ko2ic) [reply packages/image_picker/ios/Classes/ImagePickerPlugin.m]: I move these to ```FLTImagePickerPhotoAssetUtil#saveImageWithOriginalImageData```.

**Comment by ko2ic** [comment]
@cyanglaz Please review it as it has improved.

**Comment by manhnhhdvn** [comment]
@cyanglaz   Could you review this PR? I really need this feature for my individual project.

**Comment by manhnhhdvn** [comment]
@cyanglaz Can you review this PR? I really need this feature for my project

File: packages/image_picker/ios/Classes/FLTImagePickerPhotoAssetUtil.m [hunk 4]
```diff
@@ -57,11 +76,48 @@ + (NSString *)saveImageWithMetaData:(NSDictionary *)metaData
     data = [FLTImagePickerMetaDataUtil updateMetaData:metaData toImage:data];
   }
 
+  return [self createFile:data suffix:suffix];
+}
+
++ (NSString *)saveImageWithGIFInfo:(GIFInfo)info path:(NSString *)path {
```
Comment (cyanglaz) [inline packages/image_picker/ios/Classes/FLTImagePickerPhotoAssetUtil.m]: The meta data of the gif seems not copied?

Comment (ko2ic) [reply packages/image_picker/ios/Classes/FLTImagePickerPhotoAssetUtil.m]: Certainly that was not done, so I fixed it.

**Comment by ko2ic** [comment]
@cyanglaz 
I added UnitTest in ```testGetImageMIMETypeFromImageData``` and ```testSuffixFromType```.
Please re-review.

**Comment by ko2ic** [comment]
@cyanglaz I updated the pubspect.yaml and CHANGELOG.md.

File: packages/image_picker/CHANGELOG.md [hunk 5]
```diff
@@ -1,3 +1,7 @@
+## 0.6.0+10
+
+* iOS: support GIF animation and the scaling.
```
Comment (cyanglaz) [inline packages/image_picker/CHANGELOG.md]:
```suggestion
* iOS: support picking GIF from gallery.
```

Comment (ko2ic) [reply packages/image_picker/CHANGELOG.md]: Merged suggestions.

**Comment by cyanglaz** [comment]
Please also fix the formatting by running `pub global run flutter_plugin_tools format` at the plugin/ so I can merge. Sorry about the back and force.

**Comment by ko2ic** [comment]
For some reason, CI seems to have failed. Reprocessed with empty commit.

**Comment by cyanglaz** [comment]
Thank you so much for this @ko2ic. This is really great work, now people can use GIFs in image picker on iOS! Please don't hesitate to keep contributing if you have another great idea again :)

**Pull request merged**